crates/uv/src/commands/project/run.rs

```diff
@@ -51,78 +51,118 @@ pub(crate) async fn run(
-            connectivity,
-            native_tls,
-            cache,
-            printer,
-        )
-        .await?;
 
-        // Lock and sync the environment.
-        let lock = project::lock::do_lock(
-            project.workspace(),
-            venv.interpreter(),
-            settings.as_ref().into(),
-            preview,
-            connectivity,
-            concurrency,
-            native_tls,
-            cache,
-            printer,
-        )
-        .await?;
-        project::sync::do_sync(
-            project.project_name(),
-            project.workspace().root(),
-            &venv,
-            &lock,
-            extras,
-            dev,
-            Modifications::Sufficient,
-            settings.as_ref().into(),
-            preview,
-            connectivity,
-            concurrency,
-            native_tls,
-            cache,
-            printer,
-        )
-        .await?;
+        let venv = if let Some(project) = project {
+            debug!(
+                "Discovered project `{}` at: {}",
+                project.project_name(),
+                project.workspace().root().display()
+            );
+
+            let venv = project::init_environment(
+                project.workspace(),
+                python.as_deref().map(ToolchainRequest::parse),
+                toolchain_preference,
+                connectivity,
+                native_tls,
+                cache,
+                printer,
+            )
+            .await?;
+
+            // Lock and sync the environment.
+            let lock = project::lock::do_lock(
+                project.workspace(),
+                venv.interpreter(),
+                settings.as_ref().into(),
+                preview,
+                connectivity,
+                concurrency,
+                native_tls,
+                cache,
+                printer,
+            )
+            .await?;
+            project::sync::do_sync(
+                project.project_name(),
+                project.workspace().root(),
+                &venv,
+                &lock,
+                extras,
+                dev,
+                Modifications::Sufficient,
+                settings.as_ref().into(),
+                preview,
+                connectivity,
+                concurrency,
+                native_tls,
+                cache,
+                printer,
+            )
+            .await?;
+
+            venv
+        } else {
+            debug!("No project found; searching for Python interpreter");
+
+            let client_builder = BaseClientBuilder::new()
+                .connectivity(connectivity)
+                .native_tls(native_tls);
+
+            let toolchain = Toolchain::find_or_fetch(
+                python.as_deref().map(ToolchainRequest::parse),
+                EnvironmentPreference::Any,
+                toolchain_preference,
+                client_builder,
+                cache,
+            )
+            .await?;
```

Comment:
I figured `find_or_fetch` makes sense here...?

Reply:
We `find_or_fetch` lower down too. We should only fetch a toolchain if we want to create an environment. I'll need to look at the file outside the diff to see what makes sense.

Reply:
We don't install anything into this environment so it may not matter.

Reply:
If you're just looking for an environment to run a command in, `PythonEnvironment::find` is probably more appropriate? Later, if we need to install something, we need to create a new environment anyway? I'll check out the PR in a second though and read through.

Reply:
I took a look at this and made the following change. However, this comes with some downsides as we now run interpreter discovery twice when we need to fetch a toolchain — this probably isn't worth it. I think what you have is good, just sharing what I looked into for additional context. We should probably do some refactoring after this because the `find_or_fetch` in ephemeral environment construction is now _only_ for the isolated case but that's not obvious.

```diff
diff --git a/crates/uv/src/commands/project/run.rs b/crates/uv/src/commands/project/run.rs
index 368f7478..da0f9dde 100644
--- a/crates/uv/src/commands/project/run.rs
+++ b/crates/uv/src/commands/project/run.rs
@@ -14,7 +14,8 @@ use uv_distribution::{ProjectWorkspace, Workspace, WorkspaceError};
 use uv_normalize::PackageName;
 use uv_requirements::RequirementsSource;
 use uv_toolchain::{
-    EnvironmentPreference, PythonEnvironment, Toolchain, ToolchainPreference, ToolchainRequest,
+    EnvironmentPreference, Error as ToolchainError, PythonEnvironment, Toolchain,
+    ToolchainPreference, ToolchainRequest,
 };
 use uv_warnings::warn_user;
 
@@ -68,7 +69,7 @@ pub(crate) async fn run(
             }
         };
 
-        let venv = if let Some(project) = project {
+        if let Some(project) = project {
             debug!(
                 "Discovered project `{}` at: {}",
                 project.project_name(),
@@ -117,27 +118,25 @@ pub(crate) async fn run(
             )
             .await?;
 
-            venv
+            Some(venv)
         } else {
             debug!("No project found; searching for Python interpreter");
 
-            let client_builder = BaseClientBuilder::new()
-                .connectivity(connectivity)
-                .native_tls(native_tls);
-
-            let toolchain = Toolchain::find_or_fetch(
-                python.as_deref().map(ToolchainRequest::parse),
+            match PythonEnvironment::find(
+                &python
+                    .as_deref()
+                    .map(ToolchainRequest::parse)
+                    .unwrap_or_default(),
+                // No opt-in is required for system environments, since we are not mutating it
                 EnvironmentPreference::Any,
-                toolchain_preference,
-                client_builder,
                 cache,
-            )
-            .await?;
-
-            PythonEnvironment::from_toolchain(toolchain)
-        };
-
-        Some(venv)
+            ) {
+                Ok(env) => Some(env),
+                // If no environment is found, we'll create an environment later
+                Err(ToolchainError::NotFound(_)) => None,
+                Err(err) => return Err(err.into()),
+            }
+        }
     };
 
     if let Some(project_env) = &project_env {
```

Reply:
Another note... technically it is a no-no to construct a `PythonEnvironment` from a `find_or_fetch` call because you might accidentally end up allowing mutation of a managed toolchain environment. Since we're not mutating, it's fine here — but suggests that something isn't quite right in the structure (but the problem could be in the toolchain APIs themselves).

Reply:
For reference, we use the `PythonEnvironment` for two things: (1) `interpreter.scripts()`, and (2) site-packages.

Reply:
These are both attributes of the `Interpreter` and don't require a `PythonEnvironment` type, right?

Reply:
Assuming `target` and `prefix` is disabled, yes...

Reply:
https://github.com/astral-sh/uv/pull/4559